### PR TITLE
Closes #6909: LRC hashes injected while SaaS is visiting

### LINKED ISSUE #6909: 3.17 - LRC hashes must be injected when needed for SaaS visits
**Context**
This is a sub-task of #6894 to handle the injection of LRC hashes for SaaS visits.
The SaaS visits page with the query string: nowprocket=1&no_optimize=1&wpr_imagedimensions=1
This bypasses the injection of LRC hashes, preventing SaaS visits from generating valid LRC performance hints.

**Expected behavior**
Visits with the query string `wpr_imagedimensions=1` must maybe add the LRC data, under the same conditions of a normal visit (LRC feature is activated, and LRC perf. hints are not available yet in DB for this page), even if nowprocket=1&no_optimize=1 are set.

**Acceptance Criteria**
- Manually delete the LRC data of a page to ensure there is no LRC data for this page. Make sure the LRC feature is enabled. 
- Clear the cache and visit the page with `nowprocket=1&no_optimize=1&wpr_imagedimensions=1`. LRC hashes must be added to the page.
- Check the DB, the LRC data must have been stored.
- Clear the cache again. Visit the same page. LRC hashes must not be injected.

- Clear performance hints to trigger a warm-up. After a few minutes, LRC data must be available in DB and the rows must contain hashes. There could still be some inconsistencies (but just a few pages, not all). In which case, this is good enough for this issue and it will be followed-up with #6894

### PR DESCRIPTION
**Pull request by Miraeld**
# Description

Fixes #6909 

## Type of change

- [ ] New feature (non-breaking change which adds functionality).
- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] Enhancement (non-breaking change which improves an existing functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as before).
- [ ] Sub-task of #(issue number)
- [ ] Release

## Detailed scenario

- Manually delete the LRC data of a page to ensure there is no LRC data for this page. Make sure the LRC feature is enabled. 
- Clear the cache and visit the page with `nowprocket=1&no_optimize=1&wpr_imagedimensions=1`. LRC hashes must be added to the page.
- Check the DB, the LRC data must have been stored.
- Clear the cache again. Visit the same page. LRC hashes must not be injected.

### Documentation

We add a trigger to make sure hashes are added. 

### Risks

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

# Additional Checks
- [x] In the case of complex code, I wrote comments to explain it.
- [x] When possible, I prepared ways to observe the implemented system (logs, data, etc.).
- [x] I added error handling logic when using functions that could throw errors (HTTP/API request, filesystem, etc.)

### CODE REVIEW

**Comment by Miraeld** [comment]
Hasn't pass review yet, but I allow myself posting this screenshot, which is a result of the DB after a warm-up 
Screenshot 2024-08-28 at 11 30 28
We can see pages with hashes while it was full empty before. 
Pages that are empty now are normal. It should be empty. 
I did the following process: 

1. Manually delete the LRC data of a page to ensure there is no LRC data for this page. Make sure the LRC feature is enabled.

2. Clear the cache and visit the page with nowprocket=1&no_optimize=1&wpr_imagedimensions=1. LRC hashes must be added to the page.

3. Check the DB, the LRC data must have been stored.

4. Clear the cache again. Visit the same page. LRC hashes must not be injected.

5. Clear performance hints to trigger a warm-up. After a few minutes, LRC data must be available in DB and the rows must contain hashes. (Some inconsistencies can occur, it happened to me once or twice over 15 tries to see a page with empty result while it shouldn't).

**Comment by jeawhanlee** [comment]
Good Job Sir
What do you think about having an integration test for this filter: `rocket_critical_image_saas_visit_buffer` callback?

**Review by MathieuLamiot — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
Tested and looks OK on http://mathieu.e2e.rocketlabsqa.ovh/.
About https://github.com/wp-media/wp-rocket/issues/6894, I can't reproduce now with this version on mathieu.e2e.I triggered warm-ups several times and consistently get only one row with empty LRC hash list (contact-us, mobile version) which seems normal.

I'll keep this in review pending closing @jeawhanlee comment about tests. But then, it can be considered QA Done :)

**Comment by Miraeld** [comment]
Integrations tests have been added @jeawhanlee , can you confirm and move it to QA Ready ?

File: tests/Integration/inc/Engine/Optimization/LazyRenderContent/Frontend/Subscriber/add_hashes.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace WP_Rocket\Tests\Integration\inc\Engine\Optimization\LazyRenderContent\Frontend\Subscriber;
+
+use WP_Rocket\Tests\Integration\inc\Engine\Optimization\TestCase;
+
+/**
+ * Test class covering \WP_Rocket\Engine\CDN\Subscriber::rewrite
+ *
+ * @uses \WP_Rocket\Engine\Optimization\LazyRenderContent\Frontend\Subscriber::add_hashes()
+ * @group LazyRenderContent
```
Comment (jeawhanlee) [inline tests/Integration/inc/Engine/Optimization/LazyRenderContent/Frontend/Subscriber/add_hashes.php]:
```suggestion
 * @group PerformanceHints
```

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Review by jeawhanlee — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
Only tests changed since I manually checked the fix, so let's consider the QA Done

**Pull request merged**